base/io.jl

```diff
@@ -682,7 +682,7 @@ function read!(s::IO, a::Array{UInt8})
     return a
 end
 
-function read!(s::IO, a::Array{T}) where T
+function read!(s::IO, a::Union{Array{T}, FastContiguousSubArray{T,N,<:Array{T}} where N}) where T
     if isbitstype(T)
```

Comment:
```suggestion
    if isbitstype(T) && (a isa Array || a isa Base.FastContiguousSubArray{T,<:Any,<:Array{T}})
```

Explicitly documenting that this only works for `Union{Array, BitArray, SubArray}` wasn't sitting right with me — it's not quite right and it feels like an unnecessary restriction.  The only part in this method body that's not applicable to _all_ `AbstractArray`s is inside this branch!  So let's do it this way — we'll need to further widen the signature to `::AbstractArray{T}`, change the documentation to support all `::AbstractArray`s, add a few tests.

This is the "trickiest" change — we can't just ask `isa Union{...}` as that currently allocates the union itself to do the test.  This is the magic to get it for "free".

Reply:
Ok, I'm going to implement the other suggested changes.

My only concern about widening the signature to accept  any `AbstractArray` is that that would accept also structs like `UnitRange` that can't be set.  I guess that en error like "setindex! not defined for UnitRange" is not much different from a `MethodError`